asmcomp/selectgen.ml

```diff
@@ -780,6 +795,9 @@ method emit_expr (env:environment) exp =
             (nfail, ids, rs, e2, dbg))
           handlers
       in
+      let env =
+        (* Disable region-fusion on loops *)
+        match rec_flag with Recursive -> env' | Nonrecursive -> env in
       let env =
```

Comment:
Shouldn't the body always be non-tail? For cases like:
```
(region
   (catch
      (region (...; exit 0))
    with exit 0 ->
      ...))
```

Reply:
This is intended to catch cases like:
```
(region
  ...
  (catch
     (if ...
       (exit 0)
       (region E))
    with exit 0 -> ...)))
```
where `E` contains no other `exit 0` occurrences. This is a very common code pattern due to how pattern-matching is compiled, and it is useful to fuse regions in this case.

However, this will extend some lifetimes if code like your example occurs. I expect this to be very rare, so I'm tempted to leave it as is. (Such extension of lifetimes will also be caused by the simplify_exits optimisation in Simplif, I think)

I think a more sophisticated check might be to allow fusion of regions outside and in the body of a catch, if the inner region contains no references to that catch's handlers.

Reply:
I think it would be worth doing the more sophisticated check if it can be done easily. Similarly, I think it would be worth fixing `simplify_exits`.

Reply:
I have a fix for simplify_exits that I'll post upstream soon. I don't see a particularly easy way to do the more sophisticated check here.